swe_roe/ex2.c

```diff
@@ -0,0 +1,1224 @@
+
+  RDyVector     *normal;   /* unit normal vector                        */
+  RDyCoordinate *centroid; /* edge centroid                             */
+
+  PetscReal *length; /* length of the edge                        */
+
+} RDyEdge;
+
+typedef struct RDyMesh {
+  PetscInt dim;
+
+  PetscInt num_cells;
+  PetscInt num_cells_local;
+  PetscInt num_edges;
+  PetscInt num_vertices;
+  PetscInt num_boundary_faces;
+
+  PetscInt max_vertex_cells, max_vertex_faces;
+
+  RDyCell   cells;
+  RDyVertex vertices;
+  RDyEdge   edges;
+
+  PetscInt *closureSize, **closure, maxClosureSize;
+
+  PetscInt *nG2A;  // Mapping of global cells to application/natural cells
+
+} RDyMesh;
+
+struct _n_User {
+  MPI_Comm  comm;
+  char      filename[PETSC_MAX_PATH_LEN];
+  DM        dm;
+  PetscInt  Nt, Nx, Ny;
+  PetscReal dt, dx, dy;
+  PetscReal Lx, Ly;
+  PetscReal hu, hd;
+  PetscReal tiny_h;
+  PetscInt  dof, rank, comm_size;
+  Vec       B, localB;
+  Vec       localX;
+  PetscBool debug, save, add_building;
+  PetscInt  tstep;
+  PetscBool interpolate;
+
+  RDyMesh *mesh;
+};
+
+PetscErrorCode RDyInitialize_IntegerArray_1D(PetscInt *array_1D, PetscInt ndim_1, PetscInt init_value) {
+  PetscFunctionBegin;
+  for (PetscInt i = 0; i < ndim_1; i++) {
+    array_1D[i] = init_value;
+  }
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode RDyAllocate_IntegerArray_1D(PetscInt **array_1D, PetscInt ndim_1) {
+  PetscFunctionBegin;
+  PetscCall(RDyAlloc(ndim_1 * sizeof(PetscInt), array_1D));
+  PetscCall(RDyInitialize_IntegerArray_1D(*array_1D, ndim_1, -1));
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode RDyAllocate_RDyVector_1D(PetscInt ndim_1, RDyVector **array_1D) { return RDyAlloc(ndim_1 * sizeof(RDyVector), array_1D); }
+
+PetscErrorCode RDyAllocate_RDyCoordinate_1D(PetscInt ndim_1, RDyCoordinate **array_1D) { return RDyAlloc(ndim_1 * sizeof(RDyCoordinate), array_1D); }
+
+PetscErrorCode RDyAllocate_RealArray_1D(PetscReal **array_1D, PetscInt ndim_1) { return RDyAlloc(ndim_1 * sizeof(PetscReal), array_1D); }
+
+PetscBool IsClosureWithinBounds(PetscInt closure, PetscInt start, PetscInt end) { return (closure >= start) && (closure < end); }
+
+/// @brief Process command line options
+/// @param [in] comm A MPI commmunicator
+/// @param [inout] user A User data structure that is updated
+/// @return 0 on success, or a non-zero error code on failure
+static PetscErrorCode ProcessOptions(MPI_Comm comm, User user) {
+  PetscFunctionBegin;
+
+  user->comm   = comm;
+  user->Nx     = 4;
+  user->Ny     = 5;
+  user->dx     = 1.0;
+  user->dy     = 1.0;
+  user->hu     = 10.0;  // water depth for the upstream of dam   [m]
+  user->hd     = 5.0;   // water depth for the downstream of dam [m]
+  user->tiny_h = 1e-7;
+  user->dof    = 3;
+
+  MPI_Comm_size(user->comm, &user->comm_size);
+  MPI_Comm_rank(user->comm, &user->rank);
+
+  PetscErrorCode ierr;
+  ierr = PetscOptionsBegin(user->comm, NULL, "2D Mesh Options", "");
```

Comment:
In newer versions of PETSc, `PetscOptionsBegin` doesn't return an error code, so it should read

```
PetscOptionsBegin(user->comm, NULL, "2D Mesh Options", "");
```

The same applies for `PetscOptionsEnd`:

```
PetscOptionsEnd();
```